src/main/java/org/sagebionetworks/bridge/spring/controllers/ParticipantDataController.java

```diff
@@ -0,0 +1,154 @@
+    }
+
+    @GetMapping("/v3/users/self/data")
+    public ForwardCursorPagedResourceList<String> getAllDataForUser(@RequestParam(required = false) String offsetKey,
+                                                                    @RequestParam(required = false) String pageSize) {
+        UserSession session = getAuthenticatedAndConsentedSession();
+
+        int pageSizeInt = getIntOrDefault(pageSize, API_DEFAULT_PAGE_SIZE);
+        ForwardCursorPagedResourceList<ParticipantData> participantData = participantDataService
+                .getParticipantData(session.getId(), offsetKey, pageSizeInt);
+        List<String> identifiers = participantData.getItems().stream().map(ParticipantData::getIdentifier).collect(Collectors.toList());
+
+        return new ForwardCursorPagedResourceList<String>(identifiers, participantData.getNextPageOffsetKey());
+    }
+
+    @GetMapping("/v4/users/self/data/{identifier}")
+    public ParticipantData getDataByIdentifierForSelf(@PathVariable String identifier) {
+        UserSession session = getAuthenticatedAndConsentedSession();
+
+        return participantDataService.getParticipantDataRecord(session.getId(), identifier);
+    }
+
+    @PostMapping("/v4/users/self/data/{identifier}")
+    @ResponseStatus(HttpStatus.CREATED)
+    public StatusMessage saveDataRecordForSelf(@PathVariable String identifier) {
+        UserSession session = getAuthenticatedAndConsentedSession();
+
+        ParticipantData participantData = parseJson(ParticipantData.class);
+        participantData.setUserId(null); // set in service, but just so non future use depends on it
+
+        participantDataService.saveParticipantData(session.getId(), identifier, participantData);
+        return new StatusMessage("Participant data saved.");
+    }
+
+    @DeleteMapping("/v4/users/self/data/{identifier}")
+    public StatusMessage deleteDataByIdentifier(@PathVariable String identifier) {
+        UserSession session = getAuthenticatedAndConsentedSession();
+
+        ParticipantData participantData = participantDataService.getParticipantDataRecord(session.getId(), identifier);
+        if (participantData == null) {
+            throw new EntityNotFoundException(ParticipantData.class);
+        }
+        participantDataService.deleteParticipantDataRecord(session.getId(), identifier);
+
+        return new StatusMessage("Participant data record deleted.");
+    }
+
+    @GetMapping("/v1/apps/{appId}/participants/{userId}/data")
+    public ForwardCursorPagedResourceList<String> getAllDataForAdminWorker(@PathVariable String appId, @PathVariable String userId,
+                                                                           String offsetKey, String pageSize) {
+        getAuthenticatedSession(ADMIN, WORKER);
+
+        Account account = accountService.getAccount(AccountId.forId(appId, userId));
+        if (account == null) {
+            throw new EntityNotFoundException(Account.class);
+        }
```

Comment:
We also need to check that, if the caller is an Admin, that session.getAppId() is the same as the passed in appId. (Maybe in a helper function so we don't have to copy-paste it to all of the methods.

Reply:
Looking through the custom exceptions I see UnauthorizedException - if the appId of the Admin and the passed in appId don't match, is that the correct exception to throw?

Reply:
The convention is that we throw 404 not found when someone tries to access something outside of their app or study. This prevents people from fishing for a user or identifier in a study outside of their own.

403 Unauthorized is generally used of the caller doesn't have access to an API, for example if developer tries to call a researcher API.